README
GraphMan version 1.0
=======================

NAME:

Graph Manipulation Tool


DESCRIPTION:

Given a graph G(E,V), program allows to relocate, add, delete and remove edges(E) and vertices(V) in G. 



SYNOPSIS:

    #include<vector>
    #include<GraphMan.hpp>

    vector<int|long|unsigned|double> parent {1,1,2,1,2,3,4,6,3,2,67,67};
    vector<int|long|unsigned|double> child  {2,4,8,5,67,6,14,15,68,3,11,17};

/* Make Graph */

    /* Construction */
    Graph<int|long|unsigned|double> graph(parent,child);
    /*  OR  */
    Graph<int|long|unsigned|double> graph;
    graph.make(parent,child);


/*  Functions */

    graph.AddVertex(CreateVertex(),8)  // creates a new vertex and adds it to vertex 8
                                       //    result: 8 -- 69 edge
    
    graph.EraseVertex(4);  // it erases vertex 4 and makes an edge between 14 and 1

    graph.RemoveVertex(67);// vertex 67 and all its child nodes are relocated to graph 
                           // rotted at 0
                                       //    result: 0 -- 67 +- 11
                                                             +- 17
                                                             
    graph.RelocateVertex(5,6)          // assigns vertex 5 to vertex 6
                                             //    result:   1 +- 4 -- 14
                                                               +- 2 +- 8
                                                                    +- 3 +- 68
                                                                         +- 6 +- 15
                                                                              +- 5 
    vector<int|long|unsigned|double> p-dash;
    vector<int|long|unsigned|double> c-dash;
    
    graph.DmpGraph(p-dash,c-dash);      // creates two vectors 
                  //results: p-dash:  0	67	67	1	 1	2	2	3	 3	 4	 6	6	 8	 8	
                  //         c-dash: 67	11	17	2	14	8	3	6	68	14	15	5	69	70	
                  
    graph.GetDepthVec()      // returns the depth vector
                             //    result for 2: 1	2	 3	 3	2	3	 4	1	 3	 1
                             
    graph.GetVertexIdVec()   // returns the set of vertices (vector)
                             //    result for 2: 2	8	69	70	3	6	15	5	68	14



CONSTRUCTION AND RUNTIME COMPLEXITY:


     -


ACKNOWLEDGMENTS:
     
     -


AUTHOR:

Robert Bakaric <rbakaric@irb.hr>, <bakaric@evolbio.mpg.de>

COPYRIGHT AND LICENSE:

 * Copyright 2015 Robert Bakaric
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.



ChangeLog
2015-04-03  Robert Bakaric  <rbakaric@irb.hr>
	* Created! 


src/apps/GraphMan.cpp
/*
 * GraphMan.cpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

#include<iostream>
#include<fstream>
#include<vector>
#include<string>

#include<boost/program_options.hpp>

#include<GraphMan.hpp>


namespace po = boost::program_options;
using namespace std;



void PrintAcknowledgements(){

cout <<"***************************************************************************************"<<"\n";
cout <<"                           GraphMan - Graph Manipulation tool                          "<<"\n";
cout <<"                                          by                                           "<<"\n";
cout <<"                                    Robert Bakaric                                     "<<"\n\n";
cout <<"CONTACT:                                                                               "<<"\n";
cout <<" Code written and maintained by Robert Bakaric,                                        "<<"\n";
cout <<" email: rbakaric@irb.hr , bakaric@evolbio.mpg.de                                       "<<"\n\n";
cout <<"ACKNOWLEDGEMENT:                                                                       "<<"\n";
cout <<"                                                                                       "<<"\n\n";
cout <<"LICENSE:                                                                               "<<"\n";
cout <<" The program is distributed under the GNU General Public License. You should have      "<<"\n";
cout <<" received a copy of the licence together  with this software. If not, see              "<<"\n";
cout <<" http://www.gnu.org/licenses/                                                          "<<"\n";
cout <<"***************************************************************************************"<<"\n\n\n";
}


template <typename INT, typename CHARA>
po::variables_map SetOptions(INT& argc, CHARA& argv){

    try {
        int opt;
        string version = "1.0";
        po::options_description desc("Allowed options");
        desc.add_options()
            ("help,h", "produce help message")
            ("version,v", "print version information")
            ("input-file,i", po::value< string >(), "input file")
        ;

        po::positional_options_description p;
        p.add("input-file,i", -1);
        
        po::variables_map vm;
        po::store(po::command_line_parser(argc, argv).
                  options(desc).positional(p).run(), vm);
        po::notify(vm);
    
        if (vm.count("help")) {
            cout << "Usage: ./program [options]\n\n";
            PrintAcknowledgements();
            cout << desc;
            exit(0);
        }else if (vm.count("version")) {
            cout << "Program version:  " << version << "\n";
            exit(0);
        }
        if (!vm.count("input-file")){
            cout << "Input file is not defined \n";
            exit(0);
        }
        return vm;    
    }
    catch(std::exception& e)
    {
        cout << e.what() << "\n";
        exit(0);
    }    
}

template<typename T>
ostream& operator<<(ostream& os, const vector<T>& v)
{
    copy(v.begin(), v.end(), ostream_iterator<T>(os, "\t")); 
    return os;
}


template <typename Tint>
void ReadGraph(const string & file, vector<Tint>& parent, vector<Tint>& child){
  
  fstream fs;
  fs.open (file.c_str(), ios::in);
  if ( !fs.is_open()){
    cerr << "Cannot open file: "<<file <<"! Line:" << __LINE__<< endl;
    exit(-1);
  }

  int i = 0; 
  while( !fs.eof()) {
    if(child.size() == i){
       child.resize((child.size()+1)*2);
       parent.resize((parent.size()+1)*2);
    }
    fs >> child[i];
    fs >> parent[i++] ;
  }  
 
  parent.resize(i-1);
  child.resize(i-1);
  fs.close();
}



int main (int argc, char** argv){
  
/* Get cmd Options */  
  po::variables_map arg;
  arg = SetOptions(argc,argv);

  string file =  arg["input-file"].as<string>();

  vector<int> parent;
  vector<int> child;

/* Read Graph */  
  ReadGraph(file, parent, child);
  
/* Construct Graph */
  Graph<int> garph(parent,child);
  
  cout << "Graph:\n";
  cout << parent << endl;
  cout << child << endl;

  cout << "\nChanges:\n";

  int  x = garph.CreateVertex(); 
  int  y = garph.CreateVertex();

  cout << "CreateVertex:" <<  x << endl;
  cout << "CreateVertex:" <<  y << endl;

  cout << "AddVertex("<< x <<",8)\n";
  cout << "AddVertex("<< y <<",8)\n";
  garph.AddVertex(x,8);
  garph.AddVertex(y,8);

  cout << "EraseVertex(4)\n";
  garph.EraseVertex(4);

  cout << "RemoveVertex(67)\n";
  garph.RemoveVertex(67);

  cout << "\nTraversal:\n";
  garph.DepthFirstTrav(1,0,1);
  vector<int>a(garph.GetDepthVec());
  vector<int>b(garph.GetVerticesVec());
  cout << a << endl;
  cout << b << endl;

  cout << "\nNew Graph:\n";
  vector<int> p;
  vector<int> c;
  garph.DmpGraph(p,c);
  cout << p << endl;
  cout << c << endl;

  return 0;
}
